### DISCUSSION TITLE
Block Bad Bots

### DISCUSSION THREAD

**Discussion by rserafim** [vercel/community#74] [Help] [answered]
I'm trying to block some bots, but by robots.txt it's not blocking. Another simple method would be using .htaccess but Vercel doesn't have it.

Any simple way to block?

So it doesn't work.
`User-agent: BLEXBot
Disallow: /
User-agent: BLEXBot Crawler
Disallow: /
User-agent: BLEXBot/1.0
Disallow: /
User Agent: Mozilla/5.0 (compatible; BLEXBot/1.0; +http://webmeup-crawler.com/)
Disallow: /`

**Comment by Pieparker** [comment]
Hey there! There's a couple of options here: 
- make use of a `vercel.json` file as per [our .htaccess alternatives article](https://vercel.com/support/articles/does-vercel-support-htaccess-files#configuration-in-vercel) (or a `next.config.js` file, if you're using Next.js)
- create and deploy an [edge function](https://vercel.com/docs/concepts/functions/edge-functions) to mitigate against the UA

The `vercel.json` approach using [redirect rules](https://vercel.com/docs/cli#project-configuration/redirects) is arguably quicker however both options are viable!

**Reply by rserafim** [reply]
Thanks for the answer.

But where would I do a redirect? To page 401?

Was it something like that?:
`{
      "source": "/(.*)",
      "has": [
        {
          "type": "header",
          "key": "???"
        }
      ],
      "destination": "/401",
      "permanent": false
    }`

In the key what would I put in this User Agent?
Extracted from Vercel logs => User Agent:Mozilla/5.0 (compatible; BLEXBot/1.0; +http://webmeup-crawler.com/)

**Reply by Pieparker** [reply]
You could redirect to a 0 bytes (blank page) on your site or perform some blackhole routing (e.g. redirect to a URL that doesn't exist).

[Here's an example](https://github.com/vercel/next.js/discussions/21413#discussioncomment-1253882) of how the rule could look. Do be aware that this specific example is captured within a `next.config.js` file which is preferred over a `vercel.json` file, if you're using Next.js!

**Reply by rserafim** [reply]
My vercel.json file
![image](https://user-images.githubusercontent.com/48859044/148690883-be61c848-8038-454d-8a1e-e549750ccc3d.png)


Still continues (Vercel logs)
![image](https://user-images.githubusercontent.com/48859044/148690866-c49d27d9-cccc-489f-aeee-d6ef00693c8f.png)

**Reply by Pieparker** [reply]
Can you try in a `next.config.js` file?

**Reply by rserafim** [reply]
unsuccessfully! =(

![image](https://user-images.githubusercontent.com/48859044/148707698-ba537a9d-c643-4de5-a69e-7f58edd82d61.png)

![image](https://user-images.githubusercontent.com/48859044/148707722-1ebea845-6ee2-43c0-b2fb-461eb8108bc2.png)

**Reply by Pieparker** [reply]
Please share the deployment ID or URL and we can take a look at your specific project and setup.

**Reply by Pieparker** [reply]
I can see that the rule has been deployed to the project; perhaps try placing it above the `sao-paulo-sp` rule.

You can always change the (`key`/`value`) pair to test the rule with a different condition too (instead of the header rule).

**Reply by rserafim** [reply]
unsuccessfully again...

ID deploy = 5l1tfhcy8

![image](https://user-images.githubusercontent.com/48859044/148844495-e4c2c0cf-d5e2-4e02-b9b7-9802e7f72c2a.png)

vercel.json
![image](https://user-images.githubusercontent.com/48859044/148844640-9e85cf26-b1d6-4399-abbd-3920c0d3341d.png)

**Reply by Pieparker** [reply]
I've checked the cause of this error and it often occurs as a result of bad regex. Please demo the redirect on a small reproduction (perhaps using a different `key`/`value` rather than the header before attempting to integrate it to your existing project.

It's always easier to prove a concept on the smallest reproduction.

**Reply by rserafim** [reply]
Hi, I managed to come back now with this theme.

I followed your advice and made a new simple repo and applied a Header to test.

As you can see in the print it recognizes, but does not redirect. I'm wrong where?

repo: https://github.com/rserafim/headers-nextjs
![image](https://user-images.githubusercontent.com/48859044/152371317-2d27780f-e3c5-43f4-a301-86c91ec8bab2.png)

**Reply by Pieparker** [reply]
I don't believe it will be possible to redirect based on a header that you're setting in the same request. You should try redirecting based on something that's already set!

**Reply by rishi-raj-jain** [reply]
@Pieparker @leerob

The below fails to work. Can you help me? 

```
{
    "rewrites": [
        {
            "source": "/:path*",
            "has": [
                {
                    "type": "header",
                    "key": "User-Agent",
                    "value": "/bot/i"
                }
            ],
            "destination": "https://vercel.com"
        }
    ]
}
```

To reproduce:

```bash
mkdir test
npm init -y
touch vercel.json (then put the content as above)
vercel deploy --prod
```

**Reply by rishi-raj-jain** [reply]
I just checked after deploying the same code, but redirects instead of rewrites. Any particular / known thing why rewrites won't work in this case?

**Reply by rishi-raj-jain** [reply]
Oh okay, found the right fix.

https://github.com/heyxyz/hey/pull/4091/files#diff-a5de0fa0d80f14a44e9bdf2ed5d0cbb3a7bd5fdcacfbc1d03b5dcb57fc4943c1R13-R48

**Comment by rserafim** [comment, marked as answer]
What is the tested concept?
> When accessing the about page and if it is the same as my user agent, it should redirect to the 401 page.
I applied getServerSideProps on the ABOUT and 401 pages to track the logs on VERCEL.

How do I know my user agent?
> From Vercel's own logs (see prints)

Did I test it with any bot?
> Yes, I tested it with Woorank's bot.
Mozilla/5.0 (compatible; woorankreview/2.0; +https://www.woorank.com/)

RESULT: in none of the cases I got the redirect.

![image](https://user-images.githubusercontent.com/48859044/152446570-592131a2-2a4c-4fa2-883a-148f90fe29d6.png)

---

### test 1
using: next.config
https://headers-nextjs-ra7w9y4oi-agenciaubuntu.vercel.app/about
![image](https://user-images.githubusercontent.com/48859044/152445149-a2aed77d-a0c6-4ebe-9b32-22ca35c87d13.png)

---

### test 2
using: vercel.json
https://headers-nextjs-dhzmsyv1f-agenciaubuntu.vercel.app/about
![image](https://user-images.githubusercontent.com/48859044/152445706-8ed7ae10-cd8e-4e92-a07d-fff70b3b207a.png)

**Reply by Pieparker** [reply]
Can you try using the `x-matched-path: /about` or `x-powered-by: Next.js` header? I'm not sure if it's something specific about the `User Agent` header value.

**Reply by rserafim** [reply]
RESULT: in none of the cases I got the redirect.

With x-powered-by
https://headers-nextjs-g7z8kveop-agenciaubuntu.vercel.app/
![image](https://user-images.githubusercontent.com/48859044/152547074-b090c832-9a2e-4bb2-af49-f77f4c132e37.png)

With x-matched-path
https://headers-nextjs-ks22770y4-agenciaubuntu.vercel.app/
![image](https://user-images.githubusercontent.com/48859044/152547711-2e57251d-0105-4c3a-afe1-d12c09d14692.png)

**Reply by Pieparker** [reply]
When I look at the path `https://headers-nextjs-g7z8kveop-agenciaubuntu.vercel.app/about`, we are seeing other redirect rules take precedence over the `x-powered-by` header redirect. This [may be a consequence of using the `vercel.json` file](https://github.com/vercel/vercel/issues/6123) again, and not the `next.config.js` file so I'd try using that file!

**Reply by rserafim** [reply]
RESULT: in none of the cases I got the redirect. 😢

https://headers-nextjs-col7b4lcj-agenciaubuntu.vercel.app/about
![image](https://user-images.githubusercontent.com/48859044/152611535-dc3352c0-8b21-44cd-8568-4ec9ef46a83d.png)


https://headers-nextjs-lt7lyl2wy-agenciaubuntu.vercel.app/about
![image](https://user-images.githubusercontent.com/48859044/152611835-7f272dcf-54b1-4028-940b-8df02bbfca7e.png)

**Reply by Pieparker** [reply]
GitHub is currently experiencing service degradation so I spent too long wondering why my commits weren't working this evening 😆.

However, [this example](https://github.com/vercel-support/discussions-74/blob/main/next.config.js) I can assure you works.

i.e. here in the UK, if I browse to `/abc` I am redirected to `/`.

**Reply by rserafim** [reply]
Then we found that some keys don't work (x-powered-by, x-matched-path).

In the test of x-vercel-ip-country (BR) it worked correctly.

I tested the User-Agent with the regex you used on 2 bots and it also worked.

Many thanks for your support.

**Reply by dsacramone** [reply]
Instead of headers, possible to redirect by IP address or range of IPs?

**Reply by Pieparker** [reply]
Redirecting by IP address [is certainly possible](https://www.loom.com/share/7db9d8cfd1444755ac4bb751d5be105d)!

**Reply by rishi-raj-jain** [reply]
@rserafim

> I tested the User-Agent with the regex you used on 2 bots and it also worked.

Can you share this part? I'm facing errors using User-Agent as the key.

**Reply by rishi-raj-jain** [reply]
I just checked after deploying the same code, but redirects instead of rewrites. Any particular / known thing why rewrites won't work in this case?

**Reply by rserafim** [reply]
@rishi-raj-jain

I used it that way!

https://gist.github.com/rserafim/1391e9c42876df44892e249328af96e8
